python/graphstorm/model/graph_utils.py

```diff
@@ -0,0 +1,208 @@
+    # full graph evaluation
+    barrier()
+    if model.gnn_encoder is None:
+        # Only graph aware but not GNN models
+        embeddings = compute_node_input_embeddings(data.g,
+                                                   batch_size,
+                                                   model.node_input_encoder,
+                                                   task_tracker=task_tracker,
+                                                   feat_field=data.node_feat_field)
+        model.eval()
+    elif model.node_input_encoder.require_cache_embed():
+        # If the input encoder has heavy computation, we should compute
+        # the embeddings and cache them.
+        input_embeds = compute_node_input_embeddings(data.g,
+                                                     batch_size,
+                                                     model.node_input_encoder,
+                                                     task_tracker=task_tracker,
+                                                     feat_field=data.node_feat_field)
+        model.eval()
+        device = model.gnn_encoder.device
+        def get_input_embeds(input_nodes):
+            if not isinstance(input_nodes, dict):
+                assert len(data.g.ntypes) == 1
+                input_nodes = {data.g.ntypes[0]: input_nodes}
+            res = {}
+            # If the input node layer doesn't generate embeddings for a node type,
+            # we ignore it. This behavior is the same as reading node features below.
+            for ntype, ids in input_nodes.items():
+                if ntype in input_embeds:
+                    res[ntype] = input_embeds[ntype][ids].to(device)
+            return res
+        embeddings = model.gnn_encoder.dist_inference(data.g, get_input_embeds,
+                                                    batch_size, fanout, edge_mask=edge_mask,
+                                                    task_tracker=task_tracker)
+    else:
+        model.eval()
+        device = model.gnn_encoder.device
+        def get_input_embeds(input_nodes):
+            if not isinstance(input_nodes, dict):
+                assert len(data.g.ntypes) == 1
+                input_nodes = {data.g.ntypes[0]: input_nodes}
+            feats = prepare_batch_input(data.g, input_nodes, dev=device,
+                                        feat_field=data.node_feat_field)
+            return model.node_input_encoder(feats, input_nodes)
+
+        embeddings = model.gnn_encoder.dist_inference(data.g, get_input_embeds,
+                                                    batch_size, fanout, edge_mask=edge_mask,
+                                                    task_tracker=task_tracker)
+
+    if isinstance(model, GSgnnLinkPredictionModelInterface):
+        # If it is link prediction, try to normalize the node embeddings
+        # Called when model.eval()
+        embeddings = model.normalize_node_embs(embeddings)
```

Comment:
can we merge it into the model? it'll make the code more modularized. also memory efficient.

Reply:
No. 
It is very hard to do so, which may need refactoring the entire inference code especially full_graph_inference.